src/common/rbac/casbin.go

```diff
@@ -53,12 +56,85 @@ e = some(where (p.eft == allow)) && !some(where (p.eft == deny))
 m = g(r.sub, p.sub) && keyMatch2(r.obj, p.obj) && (r.act == p.act || p.act == '*')
 `
 
-// keyMatch2 determines whether key1 matches the pattern of key2, its behavior most likely the builtin KeyMatch2
-// except that the match of ("/project/1/robot", "/project/1") will return false
-func keyMatch2(key1 string, key2 string) bool {
-	key2 = strings.Replace(key2, "/*", "/.*", -1)
+type regexpEntry struct {
+	data        *regexp.Regexp
+	expiratedAt *time.Time
+}
+
+func (e *regexpEntry) isExpirated() bool {
+	return e.expiratedAt != nil && e.expiratedAt.Before(time.Now())
+}
+
+type regexpStore struct {
+	lifeDuration time.Duration
+	entries      sync.Map
+}
+
+func (s *regexpStore) Get(key string, build func(string) *regexp.Regexp) *regexp.Regexp {
+	var entry *regexpEntry
+
+	value, ok := s.entries.Load(key)
+	if !ok {
+		var expiratedAt *time.Time
+		if s.lifeDuration > 0 {
+			t := time.Now().Add(s.lifeDuration)
+			expiratedAt = &t
+		}
+
+		entry = &regexpEntry{
+			data:        build(key),
+			expiratedAt: expiratedAt,
+		}
+
+		s.entries.Store(key, entry)
+	} else {
+		entry = value.(*regexpEntry)
+	}
+
+	return entry.data
+}
+
+func (s *regexpStore) Purge() {
+	var keys []interface{}
+	s.entries.Range(func(key, value interface{}) bool {
+		if entry, ok := value.(*regexpEntry); ok && entry.isExpirated() {
+			keys = append(keys, key)
+		}
+		return true
+	})
+
+	for _, key := range keys {
+		s.entries.Delete(key)
+	}
+}
+
+var (
+	store = &regexpStore{lifeDuration: time.Hour}
+)
 
+func init() {
+	startRegexpStorePurging(store, time.Hour)
```

Comment:
enlarge it? To make it simple, remvoe the expiratedAt, just force refresh the cache?

Reply:
Force to purge the cache every 24 hours now.